### PR TITLE
Server-directed Auth Flow

### PR DESCRIPTION
**Pull request by MatkovIvan**
#319
Depends to https://github.com/Azure/azure-mobile-services-xamarin-auth/pull/5

~Only Android for now, iOS coming soon.~
@dhei let me know if anything should be updated

### CODE REVIEW

File: src/Microsoft.Azure.Mobile.Client/Authentication/MobileServicePKCEAuthentication.cs [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+        protected Uri LoginUri { get; private set; }
+
+        protected Uri CallbackUri { get; private set; }
+        
+        protected string CodeVerifier { get; private set; }
+
+        protected MobileServicePKCEAuthentication(MobileServiceClient client, string provider, string uriScheme, IDictionary<string, string> parameters)
+            : base(client, provider, parameters)
+        {
+            Debug.Assert(client != null, "client should not be null.");
```
Comment (erychagov) [inline src/Microsoft.Azure.Mobile.Client/Authentication/MobileServicePKCEAuthentication.cs]: As for me it is better to throw an exeption

Comment (MatkovIvan) [inline src/Microsoft.Azure.Mobile.Client/Authentication/MobileServicePKCEAuthentication.cs]: I chosen this way to be harmonized with other similar places in this project

File: src/Microsoft.Azure.Mobile.Client/Platforms/android/Authentication/MobileServiceUIAuthentication.cs [hunk 2]
```diff
@@ -1,48 +1,55 @@
...
                     : new InvalidOperationException (message, e.Exception);
 
                 tcs.TrySetException (ex);
+                CurrentAuthenticator = null;
             };
 
-            auth.Completed += (sender, e) =>
+            CurrentAuthenticator.Completed += (sender, e) =>
             {
                 if (!e.IsAuthenticated)
```
Comment (dhei) [inline src/Microsoft.Azure.Mobile.Client/Platforms/android/Authentication/MobileServiceUIAuthentication.cs]: nitpick: Per stylecop [SA1503 ](http://stylecop.soyuz5.com/sa1503.html) it's better not to omit curly brackets in such occasions. Can you add curly brackets here?

File: src/Microsoft.Azure.Mobile.Client/Platforms/android/Authentication/MobileServiceUIAuthentication.cs [hunk 3]
```diff
@@ -1,48 +1,55 @@
...
 
-        protected override Task<string> LoginAsyncOverride()
+        protected override Task<string> GetAuthorizationCodeAsync()
         {
             var tcs = new TaskCompletionSource<string>();
+            
+            CurrentAuthenticator = new WebRedirectAuthenticator (LoginUri, CallbackUri)
+            {
+                IsUsingNativeUI = true,
+                //ShowUIErrors = false,
```
Comment (dhei) [inline src/Microsoft.Azure.Mobile.Client/Platforms/android/Authentication/MobileServiceUIAuthentication.cs]: Can you remove this line if not needed?

File: src/Microsoft.Azure.Mobile.Client/Platforms/ios/Authentication/MobileServiceUIAuthentication.cs [hunk 4]
```diff
@@ -14,32 +14,37 @@
...
         {
             var tcs = new TaskCompletionSource<string>();
 
-            var auth = new WebRedirectAuthenticator(StartUri, EndUri);
-            auth.ShowUIErrors = false;
-            auth.ClearCookiesBeforeLogin = false;
+            CurrentAuthenticator = new WebRedirectAuthenticator(LoginUri, CallbackUri)
+            {
+                IsUsingNativeUI = ObjCRuntime.Class.GetHandle("SFSafariViewController") != IntPtr.Zero,
+                //ShowUIErrors = false,
```
Comment (dhei) [inline src/Microsoft.Azure.Mobile.Client/Platforms/ios/Authentication/MobileServiceUIAuthentication.cs]: Can you remove this line if not needed?

File: src/Microsoft.Azure.Mobile.Client/Authentication/MobileServicePKCEAuthentication.cs [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+            var loginPathAndQuery = MobileServiceUrlBuilder.CombinePathAndQuery(path, loginQueryString);
+            
+            this.LoginUri = new Uri(this.Client.MobileAppUri, loginPathAndQuery);
+            if (this.Client.AlternateLoginHost != null)
+            {
+                this.LoginUri = new Uri(this.Client.AlternateLoginHost, loginPathAndQuery);
+            }
+        }
+        
+        protected sealed override async Task<string> LoginAsyncOverride()
```
Comment (dhei) [inline src/Microsoft.Azure.Mobile.Client/Authentication/MobileServicePKCEAuthentication.cs]: This `LoginAsyncOverride()` implementation looks good, I like the way that platform-specific login ui implementation for getting `authorization_code` goes into `MobileServiceUIAuthentication.GetAuthorizationCodeAsync()`. I think it's valuable information we should add to comments. Can you add some comments before `await this.GetAuthorizationCodeAsync();` to say that `GetAuthorizationCodeAsync` deals with platform-specific login ui and callback url. Add comments after it to say client sends authorization_code and code_verifier via HTTPS request to complete the PKCE flow.

File: src/Microsoft.Azure.Mobile.Client/Platforms/android/Extensions/MobileServiceClientExtensions.cs [hunk 6]
```diff
@@ -12,7 +12,7 @@ public static partial class MobileServiceClientExtensions
         /// <summary>
         /// Log a user into a Mobile Services application given a provider name.
         /// </summary>
-        /// <param name="client" type="Microsoft.WindowsAzure.MobileServices.IMobileServiceClient">
+        /// <param name="client" type="Microsoft.WindowsAzure.MobileServices.MobileServiceClient">
```
Comment (dhei) [inline src/Microsoft.Azure.Mobile.Client/Platforms/android/Extensions/MobileServiceClientExtensions.cs]: Can you explain why change IMobileServiceClient to MobileServiceClient?

**Comment by dhei** [comment]
General notes - since we're changing public APIs in this PR, the next release should bump version to v4.0.

File: src/Microsoft.Azure.Mobile.Client/Platforms/android/Extensions/MobileServiceClientExtensions.cs (hunk 6 above)
Comment (MatkovIvan) [inline src/Microsoft.Azure.Mobile.Client/Platforms/android/Extensions/MobileServiceClientExtensions.cs]: We need access to `HttpClient` [here](https://github.com/Azure/azure-mobile-apps-net-client/pull/325/files/2bbdf443bff11aa7d241e2d3d30ba49ea097a05d#diff-c96341cf8b4da026cfe7b2aa9ec1518cR70). And I seen that `MobileServiceTokenAuthentication` use the same way.

File: src/Microsoft.Azure.Mobile.Client/Authentication/MobileServicePKCEAuthentication.cs (hunk 1 above)
Comment (dhei) [inline src/Microsoft.Azure.Mobile.Client/Authentication/MobileServicePKCEAuthentication.cs]: +1 for @erychagov - we should fail and throw when uriScheme is invalid. Who's the caller of this constructor? Can you explain what's the behavior of your code here?

File: src/Microsoft.Azure.Mobile.Client/MobileServiceUrlBuilder.cs [hunk 7]
```diff
@@ -131,6 +131,27 @@ public static string CombinePaths(string path1, string path2)
...
+        /// <param name="scheme">
+        /// The URI scheme.
+        /// </param>
+        /// <param name="path">
+        /// The URI path.
+        /// </param>
+        /// <returns>
+        /// The concatenated URI scheme and path string.
+        /// </returns>
+        public static string CombileSchemeAndPath(string scheme, string path)
```
Comment (dhei) [inline src/Microsoft.Azure.Mobile.Client/MobileServiceUrlBuilder.cs]: Can you validate parameters (scheme, path) like other utility methods in this class?

File: src/Microsoft.Azure.Mobile.Client/Authentication/MobileServicePKCEAuthentication.cs [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+
+        protected abstract Task<string> GetAuthorizationCodeAsync();
+
+        private static string GetCodeVerifier()
+        {
+            var randomBytes = WinRTCrypto.CryptographicBuffer.GenerateRandom(32);
+            return Convert.ToBase64String(randomBytes);
+        }
+
+        private static string GetHash(string data)
```
Comment (dhei) [inline src/Microsoft.Azure.Mobile.Client/Authentication/MobileServicePKCEAuthentication.cs]: Can you rename this method to include `Sha256` to be more self-descriptive?

**Comment by dhei** [comment]
General feedback. Can you add validation logic in the auth flow:
(1) validate uri scheme and and throw (similar to Android/Java implementation: [link](https://github.com/Azure/azure-mobile-apps-android-client/blob/master/sdk/src/sdk/src/main/java/com/microsoft/windowsazure/mobileservices/authentication/CustomTabsLoginManager.java#L126-L137))
(2) validate authorization_code and throw (similar to Android/Java implementation: [link](https://github.com/Azure/azure-mobile-apps-android-client/blob/master/sdk/src/sdk/src/main/java/com/microsoft/windowsazure/mobileservices/authentication/CustomTabsLoginManager.java#L139-L158))

**Comment by dhei** [comment]
General questions about concurrency. 
What's the behavior of your code when multiple login flows happen at the same time? Do you let the first login flow finish and fail all other flows? What's error/exception surfaced to the caller in such scenario?

**Comment by dhei** [comment]
General feedback about compatibility.
On platforms where SafariViewController or ChromeCustomTabs are not available, we should fallback to browser. On iOS 8 or earlier, that means Safari browser. On Android, that means browser of user's choice, such as Chrome/default browser/Firefox. At least we implemented this in Android SDK and iOS SDK. Since we are reply on Xamarin.Auth to get native UI, I am not sure of Xamarin.Auth's fallback strategy. I think we should do this if Xamarin.Auth doesn't do it.

File: src/Microsoft.Azure.Mobile.Client/Platforms/android/Extensions/MobileServiceClientExtensions.cs (hunk 6 above)
Comment (dhei) [inline src/Microsoft.Azure.Mobile.Client/Platforms/android/Extensions/MobileServiceClientExtensions.cs]: Make sense.

**Comment by MatkovIvan** [comment]
@dhei `Xamarin.Auth` don't care about fallback. I implemented fallback [here](https://github.com/Azure/azure-mobile-apps-net-client/pull/325/files#diff-82e3166173107a700bf39cf70f9b44f7R37) for iOS. I'll look at this on Android.

File: src/Microsoft.Azure.Mobile.Client/Platforms/android/Authentication/MobileServiceUIAuthentication.cs (hunk 2 above)
Comment (MatkovIvan) [inline src/Microsoft.Azure.Mobile.Client/Platforms/android/Authentication/MobileServiceUIAuthentication.cs]: Found 8 places. Fixed.

**Comment by dhei** [comment]
Looks good. Just makes sure all existing auth e2e tests (client-directed flow) still work after your change.

**Comment by MatkovIvan** [comment]
Depends to https://github.com/Azure/azure-mobile-services-xamarin-auth/pull/6

File: unittest/Microsoft.WindowsAzure.MobileServices.iOS.Test/AppDelegate.cs [hunk 9]
```diff
@@ -15,7 +16,9 @@ public partial class AppDelegate : UIApplicationDelegate
 
 	    public static TestHarness Harness { get; private set; }
 
-	    static AppDelegate()
+	    public static Func<NSUrl, bool> ResumeWithURL;
+
+        static AppDelegate()
```
Comment (dhei) [inline unittest/Microsoft.WindowsAzure.MobileServices.iOS.Test/AppDelegate.cs]: fix indentation

**Comment by MatkovIvan** [comment]
Another change in `Xamarin.Auth` https://github.com/Azure/azure-mobile-services-xamarin-auth/pull/7

**Pull request merged**